crates/cargo-util/src/process_builder.rs

```diff
@@ -398,23 +496,35 @@ impl ProcessBuilder {
 mod imp {
     use super::{ProcessBuilder, ProcessError};
     use anyhow::Result;
+    use std::io;
     use std::os::unix::process::CommandExt;
 
     pub fn exec_replace(process_builder: &ProcessBuilder) -> Result<()> {
         let mut command = process_builder.build_command();
-        let error = command.exec();
+
+        let mut error = command.exec();
+        if process_builder.should_retry_with_argfile(&error) {
+            let (mut command, _argfile) = process_builder.build_command_with_argfile()?;
+            error = command.exec()
```

Comment:
While this doesn't have the race condition between the process accessing the tempfile and us closing it, should we check for an error with close?